### ISSUE TITLE
X3 lacks boolean parser for unicode char type

### ISSUE THREAD

**Issue by saki7** [boostorg/spirit#569]
I think we could provide simple typedefs in <boost/spirit/home/x3/numeric/bool.hpp>

**Comment by Kojoley** [comment]
Do you expect it to match [synoglyphs/homoglyphs](https://en.wikipedia.org/wiki/Homoglyph), [duplicate characters in Unicode](https://en.wikipedia.org/wiki/Duplicate_characters_in_Unicode) and other Unicode quirks that I am not aware of?

**Comment by saki7** [comment]
No. It seems like the potentially required workarounds to handle those cases are beyond the scope of Spirit.

For 'other Unicode quirks' you've mentioned -- I came up with normalization (NFKC, NFKD, etc.), which I'm currently pre-processing the script in my application before passing it to X3. I think we could leave it to the user.

I think it's nice to have a documentation for Unicode support in X3, which describes the real-world use-case for building/using Unicode AST. I'm using it because my app is using code point count for UTF-32 characters. Using std::u32string::iterator, I was able to reduce the string conversion. (I think this is getting off-topic; I'd love to hear more opinions about Unicode)

**Comment by djowel** [comment]
> (I think this is getting off-topic; I'd love to hear more opinions about Unicode though)

Definitely not off-topic. I'd love to see more work done in this area.

**Comment by saki7** [comment]
For my original use-case, I'll be satisfied if `U"true"` gets parsed into `true`, for instance.

I can't imagine more complex cases for now but I really like X3's recent Unicode support so I agree with @djowel. I'll try to report more issues about Unicode soon.

**Comment by saki7** [comment]
My suggestion is changing bool_policies from
```cpp
template <typename T = bool>
struct bool_policies { /* ... */ };
```

to

```cpp
template <typename CharT, typename T = bool>
struct bool_policies; // declaration only

template <typename T = bool> // partial specialization
struct bool_policies<char32_t, T> { /* ... */ };
```

(A cleaner way is to use `if constexpr` inside the body -- IIRC it's permitted to use C++17 in X3, yes?)

If we could reach consensus then I'll try to PR.

**Comment by djowel** [comment]
> (A cleaner way is to use `if constexpr` inside the body -- IIRC it's permitted to use C++17 in X3, yes?)

I'm absolutely for a move to c++17! There will have to be some minor doc changes as I recall the requirement was c++14. It's 2020 now and we have c++20. Time to move on.

**Comment by djowel** [comment]
> (A cleaner way is to use `if constexpr` inside the body -- IIRC it's permitted to use C++17 in X3, yes?)

Oh and a lot of the infrastructure code in x3 can be simplified with c++17. Your thoughts @Kojoley ?

**Comment by saki7** [comment]
Another issue that was reported and abandoned by me. Given the fact that I've now become a maintainer of this library, I guess I need to submit a PR for this.

**Comment by saki7** [comment]
blocked by #817

**Comment by saki7** [comment]
I think this task can be done by any contributors as soon as the blocking PR is resolved. Unassigning myself for now.
